torch/csrc/jit/passes/create_functional_graphs.cpp

```diff
@@ -0,0 +1,233 @@
+         ++it) {
+      Node* n = *it;
+      if (n->kind() == prim::Constant) {
+        continue;
+      }
+      *num = *num + 1;
+      for (Block* b : n->blocks()) {
+        nonConstNodes(b, num);
+      }
+    }
+  }
+
+  bool inlineIfTooSmall(Node* n) {
+    AT_ASSERT(n->kind() == prim::FunctionalGraph);
+    auto subgraph = SubgraphUtils::getSubgraph(n);
+    size_t num_modes = 0;
+    nonConstNodes(subgraph->block(), &num_modes);
+    if (num_modes < minSubgraphSize_) {
+      SubgraphUtils::unmergeSubgraph(n);
+      return true;
+    }
+    return false;
+  }
+
+  bool CreateFunctionalGraphsImpl(Block* block) {
+    /*
+    Iterate the block in reverse and create FunctionalSubgraphs.
+    When we encounter a node that isn't functional, we skip it. Otherwise,
+    we try to merge the functional node into the current functional subgraph.
+    If it can't be merged into the current functional subgraph node, then we
+    start a functional subgraph group.
+    */
+    bool changed = false;
+    std::vector<Node*> functional_graph_nodes;
+
+    Node* functional_subgraph_node =
+        graph_->createWithSubgraph(prim::FunctionalGraph)
+            ->insertBefore(block->return_node());
+    auto reverse_iter = block->nodes().reverse();
+    std::vector<Value*> graph_outputs;
+    for (auto it = reverse_iter.begin(); it != reverse_iter.end();) {
+      Node* n = *it++;
+
+      // constants get copied into the graph
+      if (n->kind() == prim::Constant || n == functional_subgraph_node) {
+        continue;
+      }
+
+      // if `n` is functional, all of its blocks will be merged into the
+      // new functional subgraph, so we only need to recurse if it is not
+      // functional
+      if (!functional_nodes_.count(n)) {
+        for (Block* b : n->blocks()) {
+          auto block_changed = CreateFunctionalGraphsImpl(b);
+          changed = block_changed && changed;
+        }
+        continue;
+      }
+
+      if (n->kind() == prim::FunctionalGraph &&
+          isEmptyFunctionalGraph(functional_subgraph_node)) {
+        functional_subgraph_node->destroy();
+        functional_subgraph_node = n;
+        continue;
+      }
+
+      changed = true;
+      if (aliasDb_->moveBeforeTopologicallyValid(n, functional_subgraph_node)) {
+        SubgraphUtils::mergeNodeIntoSubgraph(n, functional_subgraph_node);
+      } else {
+        functional_graph_nodes.emplace_back(functional_subgraph_node);
+        functional_subgraph_node =
+            graph_->createWithSubgraph(prim::FunctionalGraph)->insertAfter(n);
+        SubgraphUtils::mergeNodeIntoSubgraph(n, functional_subgraph_node);
+      }
+    }
+    functional_graph_nodes.emplace_back(functional_subgraph_node);
+
+    for (Node* functional_node : functional_graph_nodes) {
+      if (!inlineIfTooSmall(functional_node)) {
+        ConstantPooling(functional_node->g(attr::Subgraph));
+      }
+    }
+    return changed;
+  }
+
+  bool AnalyzeFunctionalSubset(Node* n) {
```

Comment:
nit: I found this name to be confusing, because "Analyze" is not well-defined (and doesn't obviously map to a bool-returning function). Maybe `isFunctional`?